pkg/controller/config-controller.go

```diff
@@ -0,0 +1,376 @@
+			return errors.Errorf("error syncing '%s': %s", key, err.Error())
+		}
+
+		c.queue.Forget(obj)
+		glog.Infof("Successfully synced '%s'", key)
+		return nil
+
+	}(obj)
+
+	if err != nil {
+		runtime.HandleError(err)
+		return true
+	}
+
+	return true
+}
+
+func (c *ConfigController) syncHandler(key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(errors.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	selector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{})
+	if err != nil {
+		return err
+	}
+
+	// Get the CDIConfig resource with this namespace/name
+	var config *cdiv1.CDIConfig
+	configs, err := c.configLister.CDIConfigs(namespace).List(selector)
+	for _, conf := range configs {
+		if conf.Name == name {
+			config = conf
+			break
+		}
+	}
+	if config == nil {
+		// The CDIConfig resource may no longer exist, in which case we stop
+		// processing.
+		runtime.HandleError(errors.Errorf("CDIConfig '%s' in work queue no longer exists", key))
+		return nil
+	}
+	var url string
+
+	routes, err := c.routeLister.List(selector)
+	if err != nil {
+		return err
+	}
+	for _, route := range routes {
+		url = getURLFromRoute(route, c.uploadProxyServiceName)
+		if url != "" {
+			break
+		}
+	}
+	if url == "" {
+		ingresses, err := c.ingressLister.List(selector)
+		if err != nil {
+			return err
+		}
+		for _, ing := range ingresses {
+			url = getURLFromIngress(ing, c.uploadProxyServiceName)
+			if url != "" {
+				break
+			}
+		}
+	}
+	//if user overrode the upload proxy url, ignore all routes and ingresses
+	if config.Spec.UploadProxyURLOverride != nil && config.Spec.UploadProxyURLOverride != config.Status.UploadProxyURL {
+		url = *config.Spec.UploadProxyURLOverride
+	}
+	err = updateCDIConfig(c.cdiClientSet, config, url)
+	if err != nil {
+		runtime.HandleError(errors.Errorf("Error updating CDI Config %s: %s", key, err))
+	}
+	return nil
+}
+
+// Run sets up ConfigController state and executes main event loop
+func (c *ConfigController) Run(threadiness int, stopCh <-chan struct{}) error {
+	defer func() {
+		c.queue.ShutDown()
+	}()
+
+	glog.V(3).Infoln("Creating CDI config")
+	if err := c.createCDIConfig(c.configName); err != nil {
+		runtime.HandleError(err)
+		return errors.Wrap(err, "Error creating CDI config")
+	}
+
+	glog.V(3).Infoln("Starting config controller Run loop")
+	if threadiness < 1 {
+		return errors.Errorf("expected >0 threads, got %d", threadiness)
+	}
+
+	if ok := cache.WaitForCacheSync(stopCh, c.ingressesSynced, c.routesSynced); !ok {
+		return errors.New("failed to wait for caches to sync")
+	}
+
+	glog.V(3).Infoln("ConfigController cache has synced")
+
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	glog.Info("Started workers")
+	<-stopCh
+	glog.Info("Shutting down workers")
+	return nil
+}
+
+func (c *ConfigController) createCDIConfig(name string) error {
```

Comment:
I think this function only has to create a empty/default CDIConfig if one doesn't exist.  Updating the object will happen in the reconcile loop.  I would try running this controller for awhile and having it pick up a route.  Then stop the controller and delete/recreate the route with a different name.  Then restart the controller.  Pretty sure the change will get picked up.

Reply:
Curious if you ran the test I described earlier?  Because you essentially have duplicate logic here and in `syncHandler`.  If the logic cannot live in a single place, I think the two pieces of code should share a function.